util/mSCGroup.py

```diff
@@ -0,0 +1,305 @@
+# ================================================================
+# A simple class for ingestion of anndata to a TileDB group.
+#
+# * The write path (anndata/10X -> TileDB) is exercised here
+#
+# * A read path (TileDB -> anndata/10X) is a WIP.
+#
+# * The class structure in this source file is likely to change but the
+#   TileDB Groups this code creates should remain relatively stable, modulo
+#   updates to the Matrix API itself and/or any errors encountered during testing.
+#
+# * The filename is mSCGroup.py not SCGroup.py since if I have `SCGroup.py` with `class SCGroup`
+#   then I can’t do `from SCGroup import SCGroup`.
+#
+# * See also desc-ann.py in this directory for helpful information to
+#   reveal the diversity/variety of HD5 files we process.
+# ================================================================
+
+import os
+
+import anndata as ad
+import numpy   as np
+import pandas  as pd
+import pyarrow as pa
+import scanpy
+import scipy
+import tiledb
+
+
+class SCGroup():
+    """ Single-cell group
+
+    Class for representing a group of TileDB arrays that consitute an `sc_group`,
+    which includes:
+    - `X` ([`AssayMatrixGroup`]): a group of one or more labeled 2D sparse arrays
+      that share the same dimensions.
+    - `obs` ([`AnnotationDataframe`]): 1D labeled array with column labels for
+      `X`
+    - `var` ([`AnnotationDataframe`]): 1D labeled array with row labels for `X`
+    """
+
+    uri: str
+    verbose: bool
+
+    # ----------------------------------------------------------------
+    def __init__(self, uri, verbose=True):
+        """
+        @description Create a new SCGroup object. The existing array group is
+          opened at the specified array `uri` if one is present, otherwise a new
+          array group is created.
+        @param uri URI of the TileDB group
+        @param verbose Print status messages
+        """
+
+        self.uri = uri
+        self.verbose = verbose
+
+        # If URI is "/something/test1" then:
+        # * obs_uri  is "/something/test1/obs"
+        # * var_uri  is "/something/test1/var"
+        # * data_uri is "/something/test1/X"
+
+        # If URI is "tiledb://namespace/s3://bucketname/something/test1" then:
+        # * obs_uri  is "tiledb://namespace/s3://bucketname/something/test1/obs"
+        # * var_uri  is "tiledb://namespace/s3://bucketname/something/test1/var"
+        # * data_uri is "tiledb://namespace/s3://bucketname/something/test1/X"
+
+    # ----------------------------------------------------------------
+    def from_h5ad(self, input_path):
+        if self.verbose:
+            print("START  SCGroup.from_h5ad %s -> %s" % (input_path, self.uri))
+
+        anndata = self.read_h5ad(input_path)
+
+        anndata = self.decategoricalize(anndata)
+
+        self.write_tiledb_group(anndata)
+
+        if self.verbose:
+            print("FINISH SCGroup.from_h5ad %s -> %s" % (input_path, self.uri))
```

Comment:
This is a really minor comment and doesn't matter that much so don't spend time changing this now. But prefer using f-strings over %-formatting in Python 3.6+ since it is easier to read.

`print(f"FINISH SCGroup.from_h5ad {input_path} -> {self.uri}")`

Reply:
Super-easy to change, will do -- thanks @nguyenv ! :)